docs/source/developer_guide/library_design.md

```diff
@@ -0,0 +1,111 @@
+Interoperability between cuspatial, `geopandas` and other data format is maintained in `cuspatial.io` package.
+
+### UnionArray Compliance
+
+As previously mentioned,
+cuspatial's `GeoColumn` is an specialization of arrow's `UnionArray`.
+A fundamental addition in data types should be implemented in `cudf` and `GeoColumn` should simply inherits its functionality.
+However,
+`UnionArray` stands distinctly from existing data types in libcudf and requires substantial effort to implement.
+In the interim,
+cuspatial developers should build `GeoColumn` complying to `UnionArray` standards.
+Such effort may be upstreamed to `cudf` when development is more underway.
+
+## GIS computation APIs
+
+Besides the data structure,
+cuspatial maintains a set of computation APIs.
+The computation APIs are organized into several modules.
+All spatial computation modules are further grouped into a `spatial` subpackage.
+Developers are encouraged to use specific names for the module of the function added in the PR.
+
+### Legacy and Modern APIs
+
+For historical reasons,
+cuspatial APIs exposes raw array inputs for users to provide raw geometry coordinate array and offsets.
+Newer APIs should accept a `GeoSeries` or `GeoDataFrame` as input.
+Developers may extract geometry offsets and coordinates via geometry accessors provided,
+such as `GeoSeries.point`, `GeoSeries.multipoint`, `GeoSeries.lines`, `GeoSeries.polygon`.
+
+## Cython Layer
+
+The lowest layer of cuspatial is its interaction with `libcusptial` via Cython.
+The Cython layer is composed of two components: C++ bindings and Cython wrappers.
+The first component consists of [`.pxd` files](https://cython.readthedocs.io/en/latest/src/tutorial/pxd_files.html),
+Cython declaration files that expose the contents of C++ header files to other Cython files.
+The second component consists of Cython wrappers for this functionality.
+These wrappers are necessary to expose this functionality to pure Python code.
+
+To interact with the column APIs in `libcuspatial`,
+developers should have basic familarities with `libcudf` objects.
+`libcudf` is built around two principal objects whose names are largely self-explanatory: `column` and `table`.
+`libcudf` also defines corresponding non-owning "view" types `column_view` and `table_view`.
+Both `libcudf` and `libcuspatial` APIs typically accept views and return owning types.
+A `cuspatial` object owns one ore more c++ owning objects,
+the lifetime of these objects are automatically managed by python's reference counting mechanism.
+
+Similar to `cudf`,
+cython wrappers involves converting a `Column` object into `column_view` object,
+calls the `libcuspatial` API,
+and reconstructs a cudf object from the c++ result.
+By the time code reaches this stage,
+the objects are assumed to be a fully legal input to the `libcuspatial` API,
+therefore the wrapper should not contain additional component besides the above.
```

Comment:
```suggestion
The lowest layer of cuspatial is its interaction with `libcuspatial` via Cython.
The Cython layer is composed of two components: C++ bindings and
Cython wrappers. The first component consists of
[`.pxd` files](https://cython.readthedocs.io/en/latest/src/tutorial/pxd_files.html),
which are Cython declaration files that expose the contents of C++ header
files to other Cython files. The second component consists of Cython
wrappers for this functionality. These wrappers are necessary to expose
this functionality to pure Python code.

To interact with the column-based APIs in `libcuspatial`, developers should
have basic familiarity with `libcudf` objects. `libcudf` is built around two
principal objects whose names are largely self-explanatory: `column` and
`table`. `libcudf` also defines corresponding non-owning "view" types
`column_view` and `table_view`. Both `libcudf` and `libcuspatial` APIs
typically accept views and return owning types. When a `cuspatial` object
owns one ore more c++ owning objects, the lifetime of these objects is
automatically managed by python's reference counting mechanism.

Similar to cuDF, Cython wrappers must convert `Column` objects into
`column_view` objects, call the `libcuspatial` API, and reconstruct a cuDF
object from the c++ result. By the time code reaches this stage, the
objects are assumed to be fully legal inputs to the `libcuspatial` API.
Therefore the wrapper should not contain additional components besides
the above.
```

> besides the above.

Which things above?

Reply:
This is meant to say the Cython wrapper should not do additional data processing steps besides the three steps above:
convert to C objects, call the API and reconstruct the result.